### PR TITLE
JAVA-1397: Handle duration as native datatype in protocol v5+

### PR DESCRIPTION
**Pull request by olim7t**
Protocol v5 is still beta, I temporarily added this (not committed) in DurationIntegrationTest to test:
```java
    @Override
    public Cluster.Builder createClusterBuilder() {
        return super.createClusterBuilder().allowBetaProtocolVersion();
    }
```

### CODE REVIEW

**Comment by olim7t** [comment]
I think we should also update DataTypeIntegrationTest and similar, looking into that.

**Comment by olim7t** [comment]
In hindsight, it's better to keep duration tests separate because the type has specific restrictions (e.g. can't be used in collections) so it can't be handled as a "regular" primitive type by DataTypeIntegrationTest.
I've added the logic to exclude it so that things don't break when v5 becomes non-beta.

File: driver-core/src/test/java/com/datastax/driver/core/DataTypeTest.java [hunk 1]
```diff
@@ -331,7 +331,7 @@ public void should_not_return_v4_types_in_all_primitive_types_with_v3() {
         }
     }
 
-    @Test(groups = "unit")
+    @Test(groups = "unit", enabled = false)
     public void should_return_same_elements_with_all_primitive_types_using_latest_protocol_version() {
         assertThat(DataType.allPrimitiveTypes(ProtocolVersion.NEWEST_SUPPORTED)).isEqualTo(DataType.allPrimitiveTypes());
```
Comment (olim7t) [inline driver-core/src/test/java/com/datastax/driver/core/DataTypeTest.java]: This fails, I'm not sure if we should adapt it or remove it altogether.

File: driver-core/src/main/java/com/datastax/driver/core/DataType.java [hunk 2]
```diff
@@ -61,16 +60,18 @@ public boolean isCompatibleWith(Name that) {
...
         SET(34),
         MAP(33),
         CUSTOM(0),
         UDT(48, ProtocolVersion.V3),
-        TUPLE(49, ProtocolVersion.V3),
-        SMALLINT(19, ProtocolVersion.V4),
-        TINYINT(20, ProtocolVersion.V4),
-        DATE(17, ProtocolVersion.V4),
-        TIME(18, ProtocolVersion.V4);
+        TUPLE(49, ProtocolVersion.V3);
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/DataType.java:72]: If you re-order by protocolId, why not fully re-order?

File: driver-core/src/main/java/com/datastax/driver/core/DataType.java [hunk 3]
```diff
@@ -593,7 +598,10 @@ public boolean isCollection() {
 
             @Override
             public boolean apply(DataType dataType) {
-                return protocolVersion.compareTo(dataType.getName().minProtocolVersion) >= 0;
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/DataType.java]: This method is only used in tests, therefore I would suggest that we move it to `TestUtils`, wdyt? To buy us some time, I will push a commit to do that myself, but feel free to revert if you disagree.

File: driver-core/src/main/java/com/datastax/driver/core/DataType.java [hunk 4]
```diff
@@ -593,7 +598,10 @@ public boolean isCollection() {
...
+                // Duration is handled separately in DurationIntegrationTest, because it has specific restrictions (e.g.
+                // not allowed in collections).
+                return dataType.getName() != Name.DURATION &&
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/DataType.java]: Also, I think the "base" method should return all types for a given version, and let individual tests filter out undesired ones, like in `DataTypeTest.exclude(DataType)`.

File: driver-core/src/test/java/com/datastax/driver/core/DataTypeTest.java (hunk 1 above)
Comment (adutra) [reply driver-core/src/test/java/com/datastax/driver/core/DataTypeTest.java]: I vote for removing it. The method under test – `DataType.allPrimitiveTypes(ProtocolVersion)` – is itself a test method and the assumption – `NEWEST_SUPPORTED` should contain all known primitive types – is simply not true anymore.

File: driver-core/src/main/java/com/datastax/driver/core/DataType.java [hunk 5]
```diff
@@ -775,24 +783,4 @@ public String toString() {
         }
     }
 
-    /**
-     * The Duration type, introduced in Cassandra 3.10.
-     * <p/>
-     * This class is a singleton; to obtain its unique instance,
-     * call {@link #duration()}.
-     */
-    public static class DurationType extends CustomType {
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/DataType.java:784]: Pushed a small commit to delete this.

**Review by adutra — APPROVED** [review]
Approving my own changes \o/

File: driver-core/src/main/java/com/datastax/driver/core/DataType.java [hunk 6]
```diff
@@ -579,26 +582,6 @@ public boolean isCollection() {
...
-     * Returns a set of all primitive types supported by the given protocolVersion.
-     * <p>
-     * Primitive types are defined as the types that don't have type arguments
-     * (that is excluding lists, sets, and maps, tuples and udts).
-     * </p>
-     *
-     * @param protocolVersion protocol version to get types for.
-     * @return returns a set of all the primitive types for the given protocolVersion.
-     */
-    static Set<DataType> allPrimitiveTypes(final ProtocolVersion protocolVersion) {
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/DataType.java:591]: Moved to TestUtils as per my own suggestion.

File: driver-core/src/main/java/com/datastax/driver/core/DataType.java (hunk 2 above)
Comment (olim7t) [reply driver-core/src/main/java/com/datastax/driver/core/DataType.java:72]: Right, I missed a few of them.

File: driver-core/src/test/java/com/datastax/driver/core/DataTypeTest.java (hunk 1 above)
Comment (olim7t) [reply driver-core/src/test/java/com/datastax/driver/core/DataTypeTest.java]: `should_not_return_v4_types_in_all_primitive_types_with_v3` (right above) could be deleted too.

File: driver-core/src/test/java/com/datastax/driver/core/DataTypeTest.java [hunk 7]
```diff
@@ -317,22 +317,4 @@ public void serializeDeserializeCollectionsTest(ProtocolVersion version) {
...
-        // Ensure all values are <= V3.
-        for (DataType dataType : dataTypes) {
-            assertThat(dataType.getName().minProtocolVersion).isLessThanOrEqualTo(ProtocolVersion.V3);
-        }
-    }
-
-    @Test(groups = "unit")
-    public void should_return_same_elements_with_all_primitive_types_using_latest_protocol_version() {
-        assertThat(DataType.allPrimitiveTypes(ProtocolVersion.NEWEST_SUPPORTED)).isEqualTo(DataType.allPrimitiveTypes());
-    }
```
Comment (olim7t) [inline driver-core/src/test/java/com/datastax/driver/core/DataTypeTest.java:337]: Removed these two tests, as they cover a method that has been moved to TestUtils.

**Review by tolbertam — APPROVED** [review]
looks good to me! 👍

**Pull request merged**